src/engine.ts

```diff
@@ -0,0 +1,207 @@
+};
+
+export interface Tree {
+  key: string;
+  children: Node[];
+}
+
+export interface Node {
+  key: string;
+  start: number;
+  end: number;
+  tag: string;
+  text?: string;
+  children: Node[];
+}
+
+class TextNode implements Node {
+  key: string;
+  start: number;
+  end: number;
+  tag: string;
+  text?: string;
+  children: Node[];
+
+  constructor(start: number, end: number, text: string) {
+    this.key = uuid();
+    this.start = start;
+    this.end = end;
+    this.tag = ElementKind[ElementKind.span];
+    this.text = text;
+    this.children = [];
+  }
+}
+
+export interface RichTextSpan {
+  start: number;
+  end: number;
+  type: string;
+  text: string;
+}
+
+export interface RichTextBlock {
+  text: string;
+  spans: RichTextSpan[];
+}
+
+export interface Group {
+  elected: Node;
+  others: Node[];
+}
+
+interface SlicedNode {
+  inner: Node;
+  outer?: Node;
+}
+
+function sliceNode(text: string, elected: Node, node: Node): SlicedNode {
+  if (node.start < elected.start) {
+    return {
+      inner: R.merge(node, {
+        key: uuid(),
+        start: elected.start,
+        text: text.slice(elected.start, node.end),
+      }),
+      outer: R.merge(node, {
+        key: uuid(),
+        end : elected.start,
+        text: text.slice(node.start, elected.start),
+      }),
+    };
+  } else if (node.end > elected.end) {
+      return {
+        inner: R.merge(node, {
+          key: uuid(),
+          end: elected.end,
+          text: text.slice(node.start, elected.end),
+        }),
+        outer: R.merge(node, {
+          key: uuid(),
+          start: elected.end,
+          text: text.slice(elected.end, node.end),
+        }),
+      };
+  } else {
+    return {
+      inner: node,
+    };
+  }
+}
+
+function partitionGroup(text: string, group: Group): Node[] {
+  const partitioned = group.others.reduce(({ inner: innerAcc, outer: outerAcc }, node) => {
+    const slicedNode = sliceNode(text, group.elected, node);
+    return {
+      inner: innerAcc.concat(slicedNode.inner),
+      outer: slicedNode.outer ? outerAcc.concat(slicedNode.outer) : outerAcc,
+    };
+  }, { inner: new Array<Node>(), outer: new Array<Node>()});
+
+  const { inner, outer } = partitioned;
+  const head = R.assoc('children', buildTree(text, inner), group.elected);
+  return [head].concat(buildTree(text, outer));
+}
+
+function groupNodes(nodes: Node[]): Node[][] {
+  const sortByStart = (nodeA: Node, nodeB: Node) => nodeA.start < nodeB.start;
+  const sortByEnd = (nodeA: Node, nodeB: Node) => nodeA.end < nodeB.end;
+  const sortedNodes = R.sortWith([sortByStart, sortByEnd], nodes);
+  return R.groupWith((nodeA, nodeB) => nodeA.end > nodeB.start, sortedNodes);
+}
+
+function electNode(candidates: Node[]): Group {
+  if (candidates.length === 0) {
+    throw new Error('Unable to elect node on empty list');    
+  } else {
+    const [elected, ...others] = R.sortBy(node => PRIORITIES[node.tag], candidates);
```

Comment:
Should refer to the old version for priorities because if PRIORITIES[node.tag] is equal between two nodes, there are additionals rules to determine which one should be elected

Reply:
Do you mean comparing the text lenght?